edgelet/hsm-rs/src/crypto.rs

```diff
@@ -194,6 +194,30 @@ fn make_certification_props(props: &CertificateProperties) -> Result<CERT_PROPS_
             unsafe { cert_properties_destroy(handle) };
             ErrorKind::CertProps
         })?;
+
+    if !props.san_entries.is_empty() {
+        let result: Result<Vec<CString>, NulError> = props
+            .san_entries
+            .iter()
+            .map(|s: &String| CString::new(s.clone()))
+            .collect();
+
```

Comment:
It may be possible to rewrite this segment as:

```rust
if !props.san_entries.is_empty() {
        let result: Vec<*const c_char> = props
            .san_entries()
            .iter()
            .map(|s| {
                CString::new(s.as_str()).map(|s| s.as_ptr()).map_err(|_| {
                    unsafe { cert_properties_destroy(handle) };
                    ErrorKind::CertProps
                })
            }).collect::<Result<Vec<*const c_char>, ErrorKind>>()?;

        let result = unsafe { set_san_entries(handle, result.as_ptr(), result.len()) };
        if result != 0 {
            unsafe { cert_properties_destroy(handle) };
            return Err(ErrorKind::CertProps)?;
        };
    }
```

Reply:
So, as discussed offline, the above code isn't safe. We'll have to keep the `CString` objects allocated inside the `map` call alive till `set_san_entries` returns.

Reply:
The compiler doesn't complain with this code however because `set_san_entries` is being called inside an `unsafe` block.

Reply:
Tricky unsafe